include/LightGBM/tree.h

```diff
@@ -423,6 +464,145 @@ inline int Tree::PredictLeafIndex(const double* feature_values) const {
+                           PathElement *parent_unique_path, double parent_zero_fraction,
+                           double parent_one_fraction, int parent_feature_index) const {
+
+  // extend the unique path
+  PathElement *unique_path = parent_unique_path + unique_depth;
+  if (unique_depth > 0) std::copy(parent_unique_path, parent_unique_path+unique_depth, unique_path);
+  ExtendPath(unique_path, unique_depth, parent_zero_fraction,
+             parent_one_fraction, parent_feature_index);
+  const int split_index = split_feature_[node];
+
+  // leaf node
+  if (node < 0) {
+    for (int i = 1; i <= unique_depth; ++i) {
+      const double w = UnwoundPathSum(unique_path, unique_depth, i);
+      const PathElement &el = unique_path[i];
+      phi[el.feature_index] += w*(el.one_fraction-el.zero_fraction)*leaf_value_[~node];
+    }
+
+  // internal node
+  } else {
+    const int hot_index = Decision(feature_values[split_index], node);
+    const int cold_index = (hot_index == left_child_[node] ? right_child_[node] : left_child_[node]);
+    const double w = data_count(node);
+    const double hot_zero_fraction = data_count(hot_index)/w;
+    const double cold_zero_fraction = data_count(cold_index)/w;
+    double incoming_zero_fraction = 1;
+    double incoming_one_fraction = 1;
+
+    // see if we have already split on this feature,
+    // if so we undo that split so we can redo it for this node
+    int path_index = 0;
+    for (; path_index <= unique_depth; ++path_index) {
+      if (unique_path[path_index].feature_index == split_index) break;
+    }
+    if (path_index != unique_depth+1) {
+      incoming_zero_fraction = unique_path[path_index].zero_fraction;
+      incoming_one_fraction = unique_path[path_index].one_fraction;
+      UnwindPath(unique_path, unique_depth, path_index);
+      unique_depth -= 1;
+    }
+
+    TreeSHAP(feature_values, phi, hot_index, unique_depth+1, unique_path,
+             hot_zero_fraction*incoming_zero_fraction, incoming_one_fraction, split_index);
+
+    TreeSHAP(feature_values, phi, cold_index, unique_depth+1, unique_path,
+             cold_zero_fraction*incoming_zero_fraction, 0, split_index);
+  }
+}
+
+inline void Tree::PredictContrib(const double* feature_values, int num_features, double *output) const {
+  output[num_features] += ExpectedValue();
+
+  // Run the recursion with preallocated space for the unique path data
+  const int max_path_len = MaxDepth()+1;
+  PathElement *unique_path_data = new PathElement[(max_path_len*(max_path_len+1))/2];
+  TreeSHAP(feature_values, output, 0, 0, unique_path_data, 1, 1, -1);
+  delete[] unique_path_data;
+}
+
+inline double Tree::ExpectedValue(int node) const {
+  if (node >= 0) {
+    const int l = left_child_[node];
+    const int r = right_child_[node];
+    return (data_count(l)*ExpectedValue(l) + data_count(r)*ExpectedValue(r))/data_count(node);
+  } else {
+    return LeafOutput(~node);
+  }
+}
+
+inline int Tree::MaxDepth(int node) const {
```

Comment:
It seems this function only be called as ```MaxDepth(0)```, so I think using a single loop over ```leaf_depth_``` is more efficient.